File: Problem_Sets/HackerRank/SimpleTextEditor/SimpleTextEditor.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;


int main() {
    /* Enter your code here. Read input from STDIN. Print output to STDOUT */   
    int queries;
    cin >> queries;
    vector<string> prevStates;
    string s = "";
    for (int i = 0; i < queries; i++) {
        int operation;
        cin >> operation;
        cin.ignore();
        if (operation == 1) {
            string w;
            cin >> w;
            prevStates.push_back(s);
            s.append(w);
        }
        else if (operation == 2) {
            int toDel;
            cin >> toDel;
            int strLength = s.length();
            prevStates.push_back(s);
            s.erase(strLength - toDel);
        }
        else if (operation == 3) {
            int toPrint;
            cin >> toPrint;
            cout << s[toPrint - 1] << endl;
        }
        else if (operation == 4) {
            s = prevStates.back();
            prevStates.pop_back();
        }
        else {
            cout << "Invalid fn\n";
        }
        // cout << "New String: " << s << endl;
    }
    return 0;
}

File: Problem_Sets/HackerRank/SimpleTextEditor/SimpleTextEditor.py
# Enter your code here. Read input from STDIN. Print output to STDOUT
def main():
    queries = int(input())

    prevStates = []
    text = ""

    for _ in range(queries):
        inp = input()
        op = int(inp.split()[0])
        if op == 1:
            w = (inp.split()[1])
            prevStates.append(text)
            text = text + w
        elif op == 2:
            toDelete = int(inp.split()[1])
            prevStates.append(text)
            text = text[0:len(text) - toDelete:1]
        elif op == 3:
            index = int(inp.split()[1])
            print(text[index - 1])
        elif op == 4:
            text = prevStates.pop()
        else:
            print("Invalid fn call. Exiting...")


if __name__ == "__main__":
    main()